File: api/server_email.py
import os
import base64
from requests import HTTPError
from flask_cors import CORS
from flask import Flask, request, jsonify, render_template_string, make_response
from email.message import EmailMessage
from dotenv import load_dotenv
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
import webbrowser

load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))

# Configurações do OAuth
SCOPES = ["https://www.googleapis.com/auth/gmail.send"]
email_from = os.getenv('EMAIL_FROM')

# Configuração do Flask
app = Flask(__name__)
CORS(app, origins=["http://localhost:3000", "https://6021-138-0-82-55.ngrok-free.app"])

# Variável global para armazenar o serviço autenticado
service = None

# Função para autenticar com o Google OAuth
def authenticate_google():
    flow = InstalledAppFlow.from_client_secrets_file('./auth/credentials.json', SCOPES)
    creds = flow.run_local_server(port=8080)
    return build('gmail', 'v1', credentials=creds)

# Função para carregar e renderizar um modelo HTML
def load_template(structure_name, variables):
    templates_path = os.path.join(os.getcwd(), "templates/")
    template_file = os.path.join(templates_path, f"{structure_name}.html")

    if not os.path.exists(template_file):
        raise ValueError(f"Modelo para estrutura '{structure_name}' não encontrado.")

    with open(template_file, "r", encoding="utf-8") as file:
        template_content = file.read()

    return render_template_string(template_content, **variables)

# Função para criar o conteúdo do e-mail
def create_email_content(structure_name, variables):
    if not variables:
        raise ValueError(f"Erro: Nenhuma variável fornecida para o template '{structure_name}'.")

    try:
        return load_template(structure_name, variables)
    except Exception as e:
        raise ValueError(f"Erro ao criar conteúdo do e-mail: {e}")


# Função de envio de e-mail usando a API do Gmail com OAuth2
def send_email_function(variables):
    global service
    if not service:
        raise ValueError("O serviço de e-mail não foi autenticado. Chame /authenticate primeiro.")

    print(f"Dados em variables: {variables}")

    try:
        structure_name = variables.get("structure_name")
        email_content = create_email_content(structure_name, variables)

        msg = EmailMessage()
        msg.set_content(email_content, subtype="html")
        msg["To"] = variables.get("to")
        msg["From"] = email_from
        msg["Subject"] = variables.get("subject")
        secure_msg = {'raw': base64.urlsafe_b64encode(msg.as_bytes()).decode()}

        message = service.users().messages().send(userId="me", body=secure_msg).execute()
        print(f'Sent message to {variables.get("to")}. ID: {message["id"]}')

        return {"status": 200, "message": "E-mail enviado com sucesso!"}

    except HTTPError as e:
        print(f'Ocorreu um erro HTTP: {e}')
        raise e

    except Exception as e:
        print(f'Ocorreu um erro não especificado: {e}')
        raise e


# Rota para envio de e-mails
@app.route("/send-email", methods=["POST"])
def send_email():
    data = request.json
    variables = data.get("variables")

    try:
        send_email_function(variables)
        return make_response(jsonify({'message': 'E-mail enviado com sucesso.'}), 200)
    except Exception as e:
        print(f"Erro ao enviar e-mail: {e}")
        return make_response(jsonify({'message': 'Erro ao enviar e-mail.', 'error': str(e)}), 500)


# Rota para autenticação OAuth
@app.route("/authenticate", methods=["GET"])
def authenticate():
    global service  # Agora 'service' será modificada corretamente
    try:
        service = authenticate_google()  # Salva o serviço autenticado
        return jsonify({"message": "Autenticação realizada com sucesso!"})
    except Exception as e:
        return jsonify({"error": f"Erro na autenticação: {e}"}), 500


# Inicia o servidor
if __name__ == "__main__":
    port = 3001
    print(f'🚀 Servidor rodando em http://localhost:{port}')
    app.run(port=port, debug=True)
    webbrowser.open(f'https://localhost:{port}/authenticate')

File: api/server_email.mjs
/* eslint-disable no-undef */
import express from 'express'
import { google } from 'googleapis'
import nodemailer from 'nodemailer'
import open from 'open'
import dotenv from 'dotenv'
import fs from 'fs'
import path from 'path'
import bodyParser from 'body-parser'

dotenv.config({ path: '../.env' })

const app = express()
const port = 3001

app.use(bodyParser.json())

// ──────────────────
// 1. CREDENCIAIS
// ──────────────────
const CREDENTIALS_PATH = path.join(process.cwd(), './auth/credentials.json')
const credentials = JSON.parse(fs.readFileSync(CREDENTIALS_PATH))
const { client_secret, client_id, redirect_uris } = credentials.web

const oAuth2Client = new google.auth.OAuth2(
  client_id,
  client_secret, // OK mesmo com PKCE em cliente confidencial
  redirect_uris[0], // ex.: http://localhost:3001/oauth2callback
)

// ──────────────────
// 2. VARIÁVEIS GLOBAIS
// ──────────────────
let accessToken = null
let refreshToken = null
let codeVerifier = null // PKCE
const SCOPES = ['https://mail.google.com/']

// ──────────────────
// 3. FUNÇÃO util p/ templates
// ──────────────────
function loadTemplate(name, variables = {}) {
  try {
    const filePath = path.join(process.cwd(), 'templates', `${name}.html`)
    if (!fs.existsSync(filePath)) throw new Error(`Template '${name}' não encontrado.`)

    let html = fs.readFileSync(filePath, 'utf8')
    for (const key in variables)
      html = html.replace(new RegExp(`{{\\s*${key}\\s*}}`, 'g'), variables[key])

    return html
  } catch (err) {
    console.error(`Erro ao carregar template '${name}':`, err)
    throw new Error(`Template '${name}' não encontrado ou erro ao processar.`)
  }
}

// ──────────────────
// 4. /authenticate  → gera URL com PKCE e abre navegador
// ──────────────────
app.get('/authenticate', async (_req, res) => {
  // gera par verificador/desafio (biblioteca ≥ 8.0)
  const { codeVerifier: cv, codeChallenge } = await oAuth2Client.generateCodeVerifierAsync()
  codeVerifier = cv // guarda na memória (um único usuário)

  const url = oAuth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: SCOPES,
    prompt: 'consent',
    code_challenge: codeChallenge,
    code_challenge_method: 'S256',
  })

  await open(url) // abre o navegador
  res.send('🔑 Navegador aberto – conclua o login na conta Google.')
})

// ──────────────────
// 5. /oauth2callback → troca o código por tokens usando o mesmo verifier
// ──────────────────
app.get('/oauth2callback', async (req, res) => {
  const code = req.query.code?.toString()
  if (!code) return res.status(400).send('Código ausente.')

  try {
    const { tokens } = await oAuth2Client.getToken({ code, codeVerifier })
    oAuth2Client.setCredentials(tokens)

    accessToken = tokens.access_token
    refreshToken = tokens.refresh_token ?? refreshToken // guarda se vier

    res.send('✅ Autenticado com sucesso! Pode fechar esta aba.')
  } catch (err) {
    console.error(err)
    res.status(500).send(`Erro ao autenticar: ${err}`)
  }
})

// ──────────────────
// 6. /send-email    → envia e-mail via Nodemailer + OAuth2
// ──────────────────
app.post('/send-email', async (req, res) => {
  const { variables } = req.body

  if (!accessToken) {
    return res.status(401).json({ message: 'Não autenticado. Acesse /authenticate primeiro.' })
  }

  try {
    const html = loadTemplate(variables.structure_name, variables)

    const tokenResult = await oAuth2Client.getAccessToken()
    accessToken = tokenResult.token

    const transporter = nodemailer.createTransport({
      service: 'gmail',
      auth: {
        type: 'OAuth2',
        user: process.env.EMAIL_FROM,
        clientId: client_id,
        clientSecret: client_secret,
        refreshToken,
        accessToken,
      },
    })

    const result = await transporter.sendMail({
      from: process.env.EMAIL_FROM,
      to: variables.to,
      subject: variables.subject,
      html,
    })
    console.log('E-mail enviado:', result.response)
    res.json({ success: true, message: 'E-mail enviado com sucesso!' })
  } catch (err) {
    console.error('Erro ao enviar e-mail:', err)
    res.status(500).json({ message: 'Erro ao enviar e-mail.', detail: err.message })
  }
})

// ──────────────────
// 7. INICIA SERVIDOR
// ──────────────────
app.listen(port, () => console.log(`🚀  http://localhost:${port}`))
